Source/Core/Core/AchievementManager.cpp

```diff
@@ -204,6 +204,65 @@ void AchievementManager::ActivateDeactivateRichPresence()
       nullptr, 0);
 }
 
+void AchievementManager::DoFrame()
+{
+  if (!m_is_game_loaded)
+    return;
+  rc_runtime_do_frame(
+      &m_runtime,
+      [](const rc_runtime_event_t* runtime_event) {
+        AchievementManager::GetInstance()->AchievementEventHandler(runtime_event);
+      },
+      [](unsigned address, unsigned num_bytes, void* ud) {
+        return AchievementManager::GetInstance()->MemoryPeeker(address, num_bytes, ud);
+      },
+      nullptr, nullptr);
+  u64 current_time = Core::System::GetInstance().GetCoreTiming().GetTicks();
+  if (current_time - m_last_ping_time > SystemTimers::GetTicksPerSecond() * 120)
+    m_queue.EmplaceItem([this] { PingRichPresence(GenerateRichPresence()); });
+}
+
+u32 AchievementManager::MemoryPeeker(u32 address, u32 num_bytes, void* ud)
+{
+  const rc_memory_regions_t* regions = rc_console_memory_regions(m_game_data.console_id);
+  for (u32 ix = 0; ix < regions->num_regions; ix++)
+  {
+    if (address >= regions->region[ix].start_address && address <= regions->region[ix].end_address)
+    {
+      address += (regions->region[ix].real_address - regions->region[ix].start_address);
+      break;
+    }
+  }
+  Core::CPUThreadGuard guard(Core::System::GetInstance());
+  switch (num_bytes)
+  {
+  case 1:
+    return Core::System::GetInstance().GetMemory().Read_U8(address + 0x80000000);
+  case 2:
+    return Core::System::GetInstance().GetMemory().Read_U16(address + 0x80000000);
+  case 4:
+    return Core::System::GetInstance().GetMemory().Read_U32(address + 0x80000000);
+  case 8:
+    return Core::System::GetInstance().GetMemory().Read_U64(address + 0x80000000);
```

Comment:
These should use `HostRead_U8` etc. The non-host versions can trigger the emulated game's exception handler for an invalid address which is probably not what you want.

Additionally, I feel like `address + 0x80000000` is a weird choice - it seems better to use the actual virtual addresses the game uses instead of offsetting everything by 0x80000000, especially since Wii games also use addresses starting with 0x90000000 for MEM2, and some games map memory in different ways (e.g. in 0x7E000000 for GameCube games that use the MMU to page data in and out of ARAM (which isn't directly mapped)).

Reply:
The 0x80000000 was a mistake that I hadn't fully processed, but I was advised by Curtiss to use Memory.Read for physical address lockups.

Reply:
Ah, my mistake, GetRAM() and then index it directly?

Reply:
I personally think that `HostRead_U8` or `HostTryReadU8` is the right approach if you want to use addresses as the game would see them. `HostTryReadU8` also lets you choose to use the normal addresses (`Effective`) or force physical addresses. Virtual addresses are generally what you'd want though.

`HostTryReadU8` (and the other similar functions) will also call `GetRAM()` as needed (or `GetEXRAM` if it's in MEM2 and on a Wii, etc.) Reimplementing that logic seems like a bad idea to me.